docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+            public virtual int BaseSize(TargetPointer methodTablePointer) { throw new NotImplementedException(); }
+        }
+    }
+
+    namespace ContractImplementation
+    {
+        // Get contract from the predefined contract database
+        static class PredefinedContracts
+        {
+            public static IContract GetContract(string name, uint version, Target target)
+            {
+                // Do some lookup and allocate an instance of the contract requested
+            }
+        }
+
+        class MethodTableContract_1 : IAlgorithmContract
+        {
+            Target _target;
+            FieldLayout _dynamicTypeIdField;
+            FieldLayout _baseSizeField;
+
+            public MethodTableContract_1(Target target) { _target = target; }
+
+            void IAlgorithmContract.Init()
+            {
+                _dynamicTypeIdField = _target.Contract.GetFieldLayout("MethodTable", "dynamicTypeId");
+                if (_dynamicTypeIdField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+                _baseSizeField = _target.Contract.GetFieldLayout("MethodTable", "baseSize");
+                if (_baseSizeField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+            }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _dynamicTyepIdField)); }
+            public virtual int BaseSize(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _baseSizeField)); }
+        }
+
+        // This is used for version 2 of the contract, where the dynamic type id is no longer present
+        class MethodTableContract_2 : IAlgorithmContract
+        {
+            Target _target;
+            FieldLayout _baseSizeField;
+
+            public MethodTableContract_2(Target target) { _target = target; }
+
+            void IAlgorithmContract.Init()
+            {
+                _baseSizeField = _target.Contract.GetFieldLayout("MethodTable", "baseSize");
+                if (_baseSizeField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+            }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { throw new NotImplementedException(); }
+
+            public virtual int BaseSize(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _baseSizeField)); }
+        }
+    }
+}
+```
+
+Within the contract algorithm specification, the intention is that all interesting api work is done by using an instance of the `Target` class.
+
+## Arrangement of contract specifications in the repo
+
+Specs shall be stored in the repo in a set of directories. `docs/design/datacontracts`
```

Comment:
Should this spec live in that directory? I do not think we need to have a spec for a spec.

Reply:
I don't know that this spec needs to be over there, but it could be. I think of the BotR as a repository for knowledge about the runtimes and in particular for giving a roadmap to the rest of the codebase which is why I put it in the BotR.

Reply:
I look at this as a .NET runtime public API spec. BORT is a documentation of non-trivial CoreCLR internals.